crypto/crypto/src/unit_tests/cryptohasher.rs

```diff
@@ -1,62 +1,84 @@
 // Copyright (c) The Libra Core Contributors
 // SPDX-License-Identifier: Apache-2.0
 
+//! Test file for the procedural macros CryptoHasher and LCSCryptoHash.
+
 use crate as libra_crypto;
 use crate::{
-    hash::{CryptoHash, CryptoHasher},
+    hash::{CryptoHash, CryptoHasher, LIBRA_HASH_SUFFIX},
     HashValue,
 };
-use libra_crypto_derive::CryptoHasher;
+use libra_crypto_derive::{CryptoHasher, LCSCryptoHash};
+use serde::{Deserialize, Serialize};
 use tiny_keccak::{Hasher, Sha3};
 
-/// This file tests the CryptoHasher derive macro defined in the crypto-derive
-/// crate. This macro, as the comments there indicate, is meant to derive an
-/// instance of CryptoHasher (see `crate::hash`) for any type, which
-/// domain-separation tag is precisely the fully-qualified name of the
-/// struct. Here, we create a struct, manually "compute" its tag and compare it
-/// to the generated one.
+// The expected use case.
+#[derive(Serialize, Deserialize, CryptoHasher, LCSCryptoHash)]
+pub struct Foo {
+    a: u64,
+    b: u32,
+}
 
-#[derive(CryptoHasher)]
-pub struct Foo {}
+// Used for testing the seed in FooHasher.
+pub struct Bar {}
 
-impl CryptoHash for Foo {
+impl CryptoHash for Bar {
     type Hasher = FooHasher;
 
     fn hash(&self) -> HashValue {
-        // we voluntarily keep the contents empty, since we just want to test
-        // the hasher prefix
         let state = Self::Hasher::default();
         state.finish()
     }
 }
 
-// workaround of crate::hash::LIBRA_HASH_SUFFIX being private
-// the const below should always be a copy of it
-const LIBRA_HASH_SUFFIX_FOR_TESTING: &[u8] = b"@@$$LIBRA$$@@";
-
 #[test]
 fn test_cryptohasher_name() {
     let mut name = "libra_crypto::unit_tests::cryptohasher::Foo"
         .as_bytes()
         .to_vec();
-    name.extend_from_slice(LIBRA_HASH_SUFFIX_FOR_TESTING);
+    name.extend_from_slice(LIBRA_HASH_SUFFIX);
 
-    let mut digest = Sha3::v256();
-    digest.update(HashValue::from_sha3_256(&name[..]).as_ref());
+    let value = Bar {};
+    let expected = {
+        let mut digest = Sha3::v256();
+        digest.update(HashValue::from_sha3_256(&name[..]).as_ref());
+        let mut hasher_bytes = [0u8; 32];
+        digest.finalize(&mut hasher_bytes);
+        hasher_bytes
+    };
+    let actual = CryptoHash::hash(&value);
+    assert_eq!(
+        &expected,
+        actual.as_ref(),
+        "\nexpected: {} actual: {}",
+        String::from_utf8_lossy(&expected),
+        String::from_utf8_lossy(actual.as_ref())
+    );
+}
+
+#[test]
+fn test_lcs_cryptohash() {
+    let mut name = "libra_crypto::unit_tests::cryptohasher::Foo"
+        .as_bytes()
+        .to_vec();
+    name.extend_from_slice(LIBRA_HASH_SUFFIX);
 
+    let value = Foo { a: 5, b: 1025 };
     let expected = {
+        let mut digest = Sha3::v256();
+        digest.update(HashValue::from_sha3_256(&name[..]).as_ref());
+        digest.update(&value.a.to_le_bytes());
+        digest.update(&value.b.to_le_bytes());
```

Comment:
nothing like a little hand implemented LCS... why not use LCS here?

Reply:
Sure, we could do that too

Reply:
done